astropy/io/registry.py

```diff
@@ -596,7 +596,9 @@ def _get_valid_format(mode, cls, path, fileobj, args, kwargs):
 
     if len(valid_formats) == 0:
         format_table_str = _get_format_table_str(cls, mode.capitalize())
-        raise IORegistryError("Format could not be identified.\n"
+        raise IORegistryError("Format could not be identified based on the"
+                              " extension '.dat', please provide a 'format'"
```

Comment:
Hmm... Is the offending format *always* `'.dat'`?

Reply:
No, it will not always be `.dat`.  The actual answer here is not entirely trivial because it comes down to the behavior of the `identify` functions that are registered by the readers/writers.  They are actually free to decide on whether a file matches based on the file path (name, which may or may not exist) or contents or any combination thereof.

My comment earlier asking if a better error message would help was not really thought through in terms of implementation or details!

Anyway, I think we can just modify this PR slightly by saying "Format could not be identified based on the file name or contents, please provide a 'format' argument."

Reply:
I tried to print extension name using `path.split(".")[-1]` but it gives an Attribute Error when we use `Table.read()`. So I am using the error message provided by @taldcroft.

Reply:
Note that in general `path.split(".")[-1]` is not a good way to find the file "extension".  `path` might be a `Path` object which has no `split`, or it might be a string that looks like `data.fits.gz`.